deepspeed/runtime/engine.py

```diff
@@ -2710,26 +2710,29 @@ def _load_checkpoint(self,
 
         self.loaded_checkpoint_dp_world_size = checkpoint['dp_world_size']
 
+        optim_checkpoint = None
         if load_module_only:
             deepspeed_states = ['module']
             if self.optimizer is not None and self.fp16_enabled():
                 self.optimizer.refresh_fp32_params()
         else:
-            if self.has_moe_layers:
-                largest_group_name = groups._get_max_expert_size_name()
-                expp_rank = groups._get_expert_parallel_rank(largest_group_name)
-                optim_load_path = self._get_optimizer_ckpt_name(load_dir, tag, expp_rank)
-                optim_checkpoint = self.checkpoint_engine.load(optim_load_path, map_location=torch.device('cpu'))
-            else:
-                optim_checkpoint = checkpoint
-
             has_zero_optimizer_state = self.zero_optimization() or self.bfloat16_enabled()
             if load_optimizer_states and self.optimizer is not None and not has_zero_optimizer_state:
+                if self.has_moe_layers:
```

Comment:
Since MoE is composable with ZeRO, I think this MoE logic should be independent of `has_zero_optimizer_state`.

Reply:
I can change the code if you insist, @tjruwase. But then it won't be equivalent in terms of effects to what it was previously, i.e. we only `load_state_dict` for `self.optimizer` if it's `not has_zero_optimizer_state`.

Reply:
Or do you mean the fact that it's returned in `client_state` regardless?

Reply:
@clumsy, apologies my initial comment was not accurate. Also, this code has evolved in a very weird way as you have observed. Please give me a bit more time to remind myself of the original intent of this code.

Reply:
By the way, what happens if you pass `load_module_only=True`?

Reply:
Only model states will be loaded, but not LR schedule, data sampler and other auxiliary states. Optimizer state will be recreated from the model parameters, @tjruwase. See [here](https://github.com/microsoft/DeepSpeed/blob/19e9a7c0284cc886d84afd7ec58721af5e045811/deepspeed/runtime/engine.py#L2713)

Reply:
I see. I thought you wanted just model states for inference, so never mind.